nexus/src/app/sagas/instance_update/mod.rs

```diff
@@ -0,0 +1,2464 @@
+                    None
+                }
+            });
+
+        // If there's an active migration, determine how to update the instance
+        // record to reflect the current migration state.
+        if let Some(ref migration) = snapshot.migration {
+            if migration.either_side_failed() {
+                // If the migration has failed, clear the instance record's
+                // migration IDs so that a new migration can begin.
+                info!(
+                    log,
+                    "instance update (migration failed): clearing migration IDs";
+                    "instance_id" => %instance_id,
+                    "migration_id" => %migration.id,
+                    "src_propolis_id" => %migration.source_propolis_id,
+                    "target_propolis_id" => %migration.target_propolis_id,
+                );
+                new_runtime.migration_id = None;
+                new_runtime.dst_propolis_id = None;
+                update_required = true;
+                // If the active VMM was destroyed, the network config must be
+                // deleted (which was determined above). Otherwise, if the
+                // migration failed but the active VMM was still there, we must
+                // still ensure the correct networking configuration
+                // exists for its current home.
+                //
+                // TODO(#3107) This is necessary even if the instance didn't move,
+                // because registering a migration target on a sled creates OPTE ports
+                // for its VNICs, and that creates new V2P mappings on that sled that
+                // place the relevant virtual IPs on the local sled. Once OPTE stops
+                // creating these mappings, this path only needs to be taken if an
+                // instance has changed sleds.
+                if destroy_active_vmm.is_none() {
+                    if let Some(ref active_vmm) = snapshot.active_vmm {
+                        info!(
+                            log,
+                            "instance update (migration failed): pointing network \
+                             config back at current VMM";
+                            "instance_id" => %instance_id,
+                            "migration_id" => %migration.id,
+                            "src_propolis_id" => %migration.source_propolis_id,
+                            "target_propolis_id" => %migration.target_propolis_id,
+                        );
+                        network_config =
+                            Some(NetworkConfigUpdate::to_vmm(active_vmm));
+                    } else {
+                        // Otherwise, the active VMM has already been destroyed,
+                        // and the target is reporting a failure because of
+                        // that. Just delete the network config.
+                    }
+                }
+            } else if migration.either_side_completed() {
```

Comment:
Yes --- and setting `deprovision = false` in this branch is by design: An active
VMM can be destroyed because it has been migrated out of successfully and has
decided to shut down. In that case, we *do not* want to deallocate the virtual
provisioning counters, because the instance is still incarnated --- it's just on
the new sled now. We only deallocate virtual provisioning counters when the
active VMM is destroyed and the instance has *not* migrated out successfully
(which would mean it's not running anywhere).

I tried to explain this in this comment:
https://github.com/oxidecomputer/omicron/blob/a97fff14fd9a84c2ef6b5ad6d85cc083c59c4320/nexus/src/app/sagas/instance_update/mod.rs#L571-L577

If there's anything I can do to make the comment clearer, I'd love suggestions! :)

Reply:
Right, that makes sense if it's the source VMM that terminates.

Just to triple-check: Is it possible for a migration to *successfully* complete, and *also* for the "destination" VMM to terminate in one update?

(I think this is a case where we'd see "migration complete", but actually *want* the deallocation to happen, right?)

Reply:
> Just to triple-check: Is it possible for a migration to successfully complete, and also for the "destination" VMM to terminate in one update?

This is possible if the guest OS shuts down immediately after migrating (but before either of the participating sled agents actually observes the migration state update).

Reply:
@hawkw I took a second look at this and think @smklein's comment is on point here--if there's a migration, and `migration.either_side_completed` is true, but the *target* VMM is now also reporting that it's Destroyed, we should take this opportunity to clean it up and remove the provisioning counters. (I don't think the resources will get leaked in the current code, because the next periodic update saga will realize what has happened and reap the now-active VMM, but we could do it on the first pass.) We'd do this on line 592 or so--after moving the target VMM ID to the active position, we need to redo the checks in the block on line 481. WDYT?

Reply:
*Oh*, yeah, I see what you mean now. Yeah, that makes sense, I'll handle that case here too.

Reply:
@gjcolombo in the interest of making sure I'm getting this logic right: if the source Propolis has told its sled-agent that the migration out has succeeded, it's _always_ getting shut down, right? And it won't say that unless the target has also agreed that it has the instance? I just want to make sure an instance should _always_ go to "no VMM" if the migration has been reported as complete by either side when the target is destroyed.

Reply:
> if the source Propolis has told its sled-agent that the migration out has succeeded, it's always getting shut down, right?

That's correct--the end-of-migration sequence is that the target sends "I'm ready" to the source, and the source responds with "OK, you take over" and then exits. At that point the source will report that migration is complete and will tear itself down (there's no way for it to wrest control from the target anymore). If the source VMM state doesn't go to Destroyed when this happens, then there's a Propolis bug (assuming the Propolis is still otherwise responding normally).

> And it won't say that unless the target has also agreed that it has the instance?

To be precise, the target needs to *receive* "OK, you take over" from the source before it will start running. That message could always be lost, so you could end up with a situation where the source thinks migration is complete and the target thinks it's failed. In that case both sides will report that their VMMs are Destroyed (and the VM won't be running *anywhere*, which is bad but (a) could have happened anyway, e.g. if its sled rebooted, and (b) is probably still better than running it in two places at once).

Note that if this happens, the two sides will permanently disagree on whether the migration succeeded or not (source will think it's Completed but target will think it's Failed).